RavenDB-7527 Refactor subscriptions to make them easier to use

ayende opened the pull request:
@maximburyak Please take a look

ayende committed:
RavenDB-7527 Refactor subscriptions to make them easier to use

ppekrol commented on src/Raven.Client/Raven.Client.csproj:
```diff
@@ -87,4 +87,9 @@
   </ItemGroup>
+  <ItemGroup>
+    <Reference Include="System.Threading.Tasks.Extensions, Version=4.1.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51">
+      <HintPath>..\..\..\..\Users\ayende\.nuget\packages\system.threading.tasks.extensions\4.3.0\lib\netstandard1.0\System.Threading.Tasks.Extensions.dll</HintPath>
```
@ayende This needs to be removed.

ayende committed:
RavenDB-7527 - Fixing client project file

ayende committed:
RavenDB-7531
- Moving subscriptions in batch mode, instead of streaming mode

ayende committed:
Merge branch 'v4.0' of https://github.com/ravendb/ravendb into v4.0

# Conflicts:
#	test/Tryouts/Program.cs

ayende committed:
RavenDB-7531 - Removing Subscribtions / AsyncSubscription distinction.

ppekrol commented on src/Raven.Client/Documents/Session/SessionOptions.cs:
```diff
@@ -1,7 +1,10 @@
     public class SessionOptions
     {
         public string Database { get; set; }
+        public RequestExecutor RequestExecuter { get; set; }
```
typo: RequestExecutor

ayende committed:
Merge branch 'v4.0' of https://github.com/ravendb/ravendb into v4.0

ayende committed:
RavenDB-7527 - Fixing typo

ayende committed:
RavenDB-7531 - Avoid warnings and explicitly handle invalid Subscription state

ayende committed:
Merge branch 'v4.0' of https://github.com/ravendb/ravendb into v4.0

ayende committed:
Prevent the client from caching the serializer from before the CustomizeJsonSerializer call

ayende committed:
Merge branch 'v4.0' of https://github.com/ravendb/ravendb into v4.0

ayende committed:
- Fixing race condition in test
- Avoiding [Fact / Theory] error for skipped test
- Allow to run tests without calling InitializeAsync

ayende committed:
Fixing tests and avoiding compilation warnings

ayende committed:
Merge branch 'v4.0' of https://github.com/ravendb/ravendb into v4.0

# Conflicts:
#	test/FastTests/Server/Documents/Versioning/VersioningReplication.cs
#	test/Tests.Infrastructure/TestBase.cs
